File: test/ts/ssh-test/cryptoTests.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as assert from 'assert';
import { suite, test, pending, slow, timeout, params } from '@testdeck/mocha';
import {
	SshAlgorithms,
	KeyPair,
	formatBuffer,
	MessageSigner,
	MessageVerifier,
} from '@microsoft/dev-tunnels-ssh';

@suite
@slow(200)
export class CryptoTests {
	@test
	@slow(5000)
	@timeout(10000)
	@params({ kexAlg: 'diffie-hellman-group14-sha256' })
	@params({ kexAlg: 'diffie-hellman-group16-sha512' })
	@params({ kexAlg: 'ecdh-sha2-nistp521' })
	@params({ kexAlg: 'ecdh-sha2-nistp384' })
	@params({ kexAlg: 'ecdh-sha2-nistp256' })
	@params.naming((p) => `keyExchange(${p.kexAlg})`)
	public async keyExchange({ kexAlg }: { kexAlg: string }) {
		const alg = Object.values(SshAlgorithms.keyExchange).find((a) => a?.name === kexAlg)!;
		assert(alg);

		const kexA = await alg.createKeyExchange();
		const kexB = await alg.createKeyExchange();

		const exchangeA = await kexA.startKeyExchange();
		const exchangeB = await kexB.startKeyExchange();

		const secretA = await kexA.decryptKeyExchange(exchangeB);
		const secretB = await kexB.decryptKeyExchange(exchangeA);

		assert(secretB.equals(secretA), 'key exchange secret');
	}

	@test
	@slow(5000)
	@timeout(10000)
	@params({ pkAlg: 'rsa-sha2-256', keySize: 1024 })
	@params({ pkAlg: 'rsa-sha2-512', keySize: 2048 })
	@params({ pkAlg: 'rsa-sha2-512', keySize: 4096 })
	@params({ pkAlg: 'ecdsa-sha2-nistp256' })
	@params({ pkAlg: 'ecdsa-sha2-nistp384' })
	@params({ pkAlg: 'ecdsa-sha2-nistp521' })
	@params.naming((p) => `signVerify(${p.pkAlg})`)
	public async signVerify({ pkAlg, keySize }: { pkAlg: string; keySize?: number }) {
		const alg = Object.values(SshAlgorithms.publicKey).find((a) => a?.name === pkAlg)!;
		assert(alg);

		const keyPair = await alg.generateKeyPair(keySize);

		const data = Buffer.from('test');
		const signer = alg.createSigner(keyPair);
		const signature = await signer.sign(data);
		assert(signature, 'signing with key pair');

		const verifier = alg.createVerifier(keyPair);
		const verified = await verifier.verify(data, signature);
		assert(verified, 'verifying with key pair');
	}

	@test
	@params({ encAlg: 'aes256-ctr' })
	@params({ encAlg: 'aes256-gcm@openssh.com' })
	@params.naming((p) => `encryptDecrypt(${p.encAlg})`)
	public async encryptDecrypt({ encAlg }: { encAlg: string }) {
		const alg = Object.values(SshAlgorithms.encryption).find((a) => a?.name === encAlg)!;
		assert(alg);

		const key = Buffer.alloc(alg.keyLength);
		const iv = Buffer.alloc(alg.blockLength);

		const random = SshAlgorithms.random;
		random.getBytes(key);
		random.getBytes(iv);

		const cipher = await alg.createCipher(true, key, iv);
		const decipher = await alg.createCipher(false, key, iv);

		const plaintext = Buffer.alloc(3 * alg.blockLength);
		plaintext.fill(9);

		const ciphertext = await cipher.transform(plaintext);
		assert.equal(ciphertext.length, plaintext.length, 'encrypted length');
		assert(!ciphertext.equals(plaintext), 'encrypted data');

		const signer = <MessageSigner>(<unknown>cipher);
		const verifier = <MessageVerifier>(<unknown>decipher);
		if (signer.authenticatedEncryption) {
			const tag = await signer.sign(plaintext);
			await verifier.verify(ciphertext, tag);
		}

		const plaintext2 = await decipher.transform(ciphertext);
		assert(plaintext2.equals(plaintext), 'decrypted data');
	}

	@test
	@params({ hmacAlg: 'hmac-sha2-256' })
	@params({ hmacAlg: 'hmac-sha2-512' })
	@params({ hmacAlg: 'hmac-sha2-256-etm@openssh.com' })
	@params({ hmacAlg: 'hmac-sha2-512-etm@openssh.com' })
	@params.naming((p) => `hmac(${p.hmacAlg})`)
	public async hmac({ hmacAlg }: { hmacAlg: string }) {
		const alg = Object.values(SshAlgorithms.hmac).find((a) => a?.name === hmacAlg)!;
		assert(alg);

		const key = Buffer.alloc(alg.keyLength);

		const random = SshAlgorithms.random;
		random.getBytes(key);

		const signer = await alg.createSigner(key);
		const verifier = await alg.createVerifier(key);

		const data = Buffer.alloc(16);
		data.fill(9);

		const signature = await signer.sign(data);
		assert.equal(signature.length, alg.digestLength, 'digest length');

		const verified = await verifier.verify(data, signature);
		assert(verified, 'verify');
	}
}

File: test/cs/Ssh.Test/CryptoTests.cs
using Microsoft.DevTunnels.Ssh.Algorithms;
using System;
using System.Linq;
using System.Reflection;
using System.Text;
using Xunit;

namespace Microsoft.DevTunnels.Ssh.Test;

public class CryptoTests
{
	[Theory]
	[InlineData(DiffieHellman.DHGroup14Sha256)]
	[InlineData(DiffieHellman.DHGroup16Sha512)]
#if SSH_ENABLE_ECDH
	[InlineData(ECDiffieHellman.EcdhNistp256)]
	[InlineData(ECDiffieHellman.EcdhNistp384)]
	[InlineData(ECDiffieHellman.EcdhNistp521)]
#endif
	public void KeyExchange(string kexAlg)
	{
		var alg = GetAlgorithmByName<KeyExchangeAlgorithm>(
			typeof(SshAlgorithms.KeyExchange), kexAlg);
		Assert.True(alg.IsAvailable, $"Algorithm not available: {kexAlg}");

		using var kexA = alg.CreateKeyExchange();
		using var kexB = alg.CreateKeyExchange();

		var exchangeA = kexA.StartKeyExchange();
		var exchangeB = kexB.StartKeyExchange();

		var secretA = kexA.DecryptKeyExchange(exchangeB);
		var secretB = kexB.DecryptKeyExchange(exchangeA);

		Assert.True(secretB.Equals(secretA));
	}

	[Theory]
	[InlineData(Rsa.RsaWithSha256, 1024)]
	[InlineData(Rsa.RsaWithSha512, 2048)]
	[InlineData(Rsa.RsaWithSha512, 4096)]
	[InlineData(ECDsa.ECDsaSha2Nistp256, null)]
	[InlineData(ECDsa.ECDsaSha2Nistp384, null)]
	[InlineData(ECDsa.ECDsaSha2Nistp521, null)]
	public void SignVerify(string pkAlg, int? keySize)
	{
		var alg = GetAlgorithmByName<PublicKeyAlgorithm>(
			typeof(SshAlgorithms.PublicKey), pkAlg);
		Assert.True(alg.IsAvailable, $"Algorithm not available: {pkAlg}");

		var keyPair = alg.GenerateKeyPair(keySize);

		Buffer data = Encoding.UTF8.GetBytes("test");
		var signer = alg.CreateSigner(keyPair);

		var signature = new Buffer(signer.DigestLength);
		signer.Sign(data, signature);

		var verifier = alg.CreateVerifier(keyPair);
		var verified = verifier.Verify(data, signature);
		Assert.True(verified);
	}

	[Theory]
	[InlineData("aes256-cbc")]
	[InlineData("aes256-ctr")]
#if SSH_ENABLE_AESGCM
	[InlineData("aes256-gcm@openssh.com")]
#endif
	public void EncryptDecrypt(string encAlg)
	{
		var alg = GetAlgorithmByName<EncryptionAlgorithm>(
			typeof(SshAlgorithms.Encryption), encAlg);
		Assert.True(alg.IsAvailable, $"Algorithm not available: {encAlg}");

		var key = new Buffer(alg.KeyLength);
		var iv = new Buffer(alg.BlockLength);

		var random = SshAlgorithms.Random;
		random.GetBytes(key);
		random.GetBytes(iv);

		var cipher = alg.CreateCipher(true, key, iv);
		var decipher = alg.CreateCipher(false, key, iv);

		var plaintext = new Buffer(3 * alg.BlockLength);
		for (int i = 0; i < plaintext.Count; i++)
		{
			plaintext[i] = (byte)i;
		}

		var ciphertext = new Buffer(plaintext.Count);
		cipher.Transform(plaintext, ciphertext);
		Assert.False(ciphertext.SequenceEqual(plaintext));

		var signer = cipher as IMessageSigner;
		var verifier = decipher as IMessageVerifier;
		if (signer?.AuthenticatedEncryption == true)
		{
			var tag = new Buffer(signer.DigestLength);
			signer.Sign(plaintext, tag);
			verifier.Verify(ciphertext, tag);
		}

		var plaintext2 = new Buffer(ciphertext.Count);
		decipher.Transform(ciphertext, plaintext2);
		Assert.True(plaintext2.SequenceEqual(plaintext));
	}

	[Theory]
	[InlineData("hmac-sha2-256")]
	[InlineData("hmac-sha2-512")]
	[InlineData("hmac-sha2-256-etm@openssh.com")]
	[InlineData("hmac-sha2-512-etm@openssh.com")]
	public void Hmac(string hmacAlg)
	{
		var alg = GetAlgorithmByName<HmacAlgorithm>(
			typeof(SshAlgorithms.Hmac), hmacAlg);
		Assert.True(alg.IsAvailable, $"Algorithm not available: {hmacAlg}");

		var key = new Buffer(alg.KeyLength);

		var random = SshAlgorithms.Random;
		random.GetBytes(key);

		var signer = alg.CreateSigner(key);
		var verifier = alg.CreateVerifier(key);

		var data = new Buffer(16);
		for (int i = 0; i < data.Count; i++)
		{
			data[i] = (byte)i;
		}

		var signature = new Buffer(signer.DigestLength);
		signer.Sign(data, signature);

		var verified = verifier.Verify(data, signature);
		Assert.True(verified);

	}

	private static T GetAlgorithmByName<T>(Type algorithmClass, string name)
		where T : SshAlgorithm
	{
		return algorithmClass.GetProperties(BindingFlags.Public | BindingFlags.Static)
			.Select((p) => p.GetValue(null))
			.Cast<T>()
			.FirstOrDefault((a) => a?.Name == name) ??
			throw new ArgumentException($"Algorithm not found: {name}");
	}
}
